File: VB/PayrollRepository.SalaryGroup.vb
Imports System.Linq.Expressions
Imports LinqKit.Extensions
Imports System.Data.Common
Imports Framework.Data
Imports Framework.Data.System.Linq.Dynamic
Imports System.Data.Entity
Imports System.Text.RegularExpressions
Imports System.Data.Objects
Imports System.Reflection

Partial Public Class PayrollRepository

#Region "SalaryGroup list"

    Public Function GetSalaryGroup(ByVal _filter As SalaryGroupDTO, ByVal PageIndex As Integer,
                                        ByVal PageSize As Integer,
                                        ByRef Total As Integer,
                                        Optional ByVal Sorts As String = "CREATED_DATE desc") As List(Of SalaryGroupDTO)

        Try
            Dim query = From p In Context.PA_SALARY_GROUP

            If _filter.CODE <> "" Then
                query = query.Where(Function(p) p.CODE.ToUpper.Contains(_filter.CODE.ToUpper))
            End If

            If _filter.NAME <> "" Then
                query = query.Where(Function(p) p.NAME.ToUpper.Contains(_filter.NAME.ToUpper))
            End If

            If _filter.EFFECT_DATE IsNot Nothing Then
                query = query.Where(Function(p) p.EFFECT_DATE = _filter.EFFECT_DATE)
            End If

            Dim lst = query.Select(Function(p) New SalaryGroupDTO With {
                                       .ID = p.ID,
                                       .CODE = p.CODE,
                                       .NAME = p.NAME,
                                       .REMARK = p.REMARK,
                                       .EFFECT_DATE = p.EFFECT_DATE,
                                       .CREATED_DATE = p.CREATED_DATE})


            lst = lst.OrderBy(Sorts)
            Total = lst.Count
            lst = lst.Skip(PageIndex * PageSize).Take(PageSize)

            Return lst.ToList

        Catch ex As Exception
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod.Name, "iPayroll")
            Throw ex
        End Try


    End Function

    Public Function GetEffectSalaryGroup() As SalaryGroupDTO
        Try
            Dim query = From p In Context.PA_SALARY_GROUP
            Where p.EFFECT_DATE <= Date.Now
                 Order By p.EFFECT_DATE Descending, p.CREATED_DATE Descending

            Dim EffectSalaryGroup = query.Select(Function(p) New SalaryGroupDTO With {
                                       .ID = p.ID,
                                       .CODE = p.CODE,
                                       .NAME = p.NAME,
                                       .REMARK = p.REMARK,
                                       .EFFECT_DATE = p.EFFECT_DATE,
                                       .CREATED_DATE = p.CREATED_DATE}).FirstOrDefault

            Return EffectSalaryGroup
        Catch ex As Exception
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod.Name, "iPayroll")
            Throw ex
        End Try


    End Function

    Public Function InsertSalaryGroup(ByVal objSalaryGroup As SalaryGroupDTO,
                                   ByVal log As UserLog, ByRef gID As Decimal) As Boolean
        Dim iCount As Integer = 0
        Dim objSalaryGroupData As New PA_SALARY_GROUP
        Try
            objSalaryGroupData.ID = Utilities.GetNextSequence(Context, Context.PA_SALARY_GROUP.EntitySet.Name)
            objSalaryGroupData.CODE = objSalaryGroup.CODE.Trim
            objSalaryGroupData.NAME = objSalaryGroup.NAME.Trim
            objSalaryGroupData.EFFECT_DATE = objSalaryGroup.EFFECT_DATE
            objSalaryGroupData.REMARK = objSalaryGroup.REMARK
            Context.PA_SALARY_GROUP.AddObject(objSalaryGroupData)
            Context.SaveChanges(log)


            Context.SaveChanges(log)
            gID = objSalaryGroupData.ID
            Return True
        Catch ex As Exception
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod.Name, "iPayroll")
            Throw ex
        End Try

    End Function

    Public Function ValidateSalaryGroup(ByVal _validate As SalaryGroupDTO)
        Dim query
        Try
            If _validate.CODE <> Nothing Then
                If _validate.ID <> 0 Then
                    query = (From p In Context.PA_SALARY_GROUP
                             Where p.CODE.ToUpper = _validate.CODE.ToUpper _
                             And p.ID <> _validate.ID).SingleOrDefault
                Else
                    query = (From p In Context.PA_SALARY_GROUP
                             Where p.CODE.ToUpper = _validate.CODE.ToUpper).FirstOrDefault
                End If
                Return (query Is Nothing)
            End If
            Return True
        Catch ex As Exception
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod.Name, "iPayroll")
            Throw ex
        End Try
    End Function

    Public Function ModifySalaryGroup(ByVal objSalaryGroup As SalaryGroupDTO,
                                   ByVal log As UserLog, ByRef gID As Decimal) As Boolean
        Dim objSalaryGroupData As New PA_SALARY_GROUP With {.ID = objSalaryGroup.ID}
        Try
            Context.PA_SALARY_GROUP.Attach(objSalaryGroupData)
            objSalaryGroupData.CODE = objSalaryGroup.CODE.Trim
            objSalaryGroupData.NAME = objSalaryGroup.NAME.Trim
            objSalaryGroupData.EFFECT_DATE = objSalaryGroup.EFFECT_DATE
            objSalaryGroupData.REMARK = objSalaryGroup.REMARK

            Context.SaveChanges(log)
            gID = objSalaryGroupData.ID
            Return True
        Catch ex As Exception
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod.Name, "iPayroll")
            Throw ex
        End Try

    End Function

    Public Function DeleteSalaryGroup(ByVal lstID As List(Of Decimal)) As Boolean
        Dim lstSalaryGroupData As List(Of PA_SALARY_GROUP)
        Try
            lstSalaryGroupData = (From p In Context.PA_SALARY_GROUP Where lstID.Contains(p.ID)).ToList

            For idx = 0 To lstSalaryGroupData.Count - 1
                For Each item In lstSalaryGroupData(idx).PA_SALARY_LEVEL
                    For Each item1 In item.PA_SALARY_RANK
                        Context.PA_SALARY_RANK.DeleteObject(item1)
                    Next
                    Context.PA_SALARY_LEVEL.DeleteObject(item)
                Next

                Context.PA_SALARY_GROUP.DeleteObject(lstSalaryGroupData(idx))
            Next
            Context.SaveChanges()
            Return True
        Catch ex As Exception
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod.Name, "iPayroll")
            Throw ex
        End Try

    End Function

#End Region


End Class

File: C#/PayrollRepository.SalaryGroup.cs

using System.Linq.Expressions;
using LinqKit.Extensions;
using System.Data.Common;
using Framework.Data;
using Framework.Data.System.Linq.Dynamic;
using System.Data.Entity;
using System.Text.RegularExpressions;
using System.Data.Objects;

public partial class PayrollRepository
{
    public List<SalaryGroupDTO> GetSalaryGroup(SalaryGroupDTO _filter, int PageIndex, int PageSize, ref int Total, string Sorts = "CREATED_DATE desc")
    {
        try
        {
            var query = from p in Context.PA_SALARY_GROUP
                        select p;

            if (_filter.CODE != "")
                query = query.Where(p => p.CODE.ToUpper.Contains(_filter.CODE.ToUpper));

            if (_filter.NAME != "")
                query = query.Where(p => p.NAME.ToUpper.Contains(_filter.NAME.ToUpper));

            if (_filter.EFFECT_DATE != null)
                query = query.Where(p => p.EFFECT_DATE == _filter.EFFECT_DATE);

            var lst = query.Select(p => new SalaryGroupDTO() { ID = p.ID, CODE = p.CODE, NAME = p.NAME, REMARK = p.REMARK, EFFECT_DATE = p.EFFECT_DATE, CREATED_DATE = p.CREATED_DATE });


            lst = lst.OrderBy(Sorts);
            Total = lst.Count;
            lst = lst.Skip(PageIndex * PageSize).Take(PageSize);

            return lst.ToList;
        }
        catch (Exception ex)
        {
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod().Name, "iPayroll");
            throw ex;
        }
    }

    public SalaryGroupDTO GetEffectSalaryGroup()
    {
        try
        {
            var query = from p in Context.PA_SALARY_GROUP
                        where p.EFFECT_DATE <= DateTime.Now
                        orderby p.EFFECT_DATE descending, p.CREATED_DATE descending
                        select p;

            var EffectSalaryGroup = query.Select(p => new SalaryGroupDTO() { ID = p.ID, CODE = p.CODE, NAME = p.NAME, REMARK = p.REMARK, EFFECT_DATE = p.EFFECT_DATE, CREATED_DATE = p.CREATED_DATE }).FirstOrDefault;

            return EffectSalaryGroup;
        }
        catch (Exception ex)
        {
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod().Name, "iPayroll");
            throw ex;
        }
    }

    public bool InsertSalaryGroup(SalaryGroupDTO objSalaryGroup, UserLog log, ref decimal gID)
    {
        int iCount = 0;
        PA_SALARY_GROUP objSalaryGroupData = new PA_SALARY_GROUP();
        try
        {
            objSalaryGroupData.ID = Utilities.GetNextSequence(Context, Context.PA_SALARY_GROUP.EntitySet.Name);
            objSalaryGroupData.CODE = objSalaryGroup.CODE.Trim;
            objSalaryGroupData.NAME = objSalaryGroup.NAME.Trim;
            objSalaryGroupData.EFFECT_DATE = objSalaryGroup.EFFECT_DATE;
            objSalaryGroupData.REMARK = objSalaryGroup.REMARK;
            Context.PA_SALARY_GROUP.AddObject(objSalaryGroupData);
            Context.SaveChanges(log);


            Context.SaveChanges(log);
            gID = objSalaryGroupData.ID;
            return true;
        }
        catch (Exception ex)
        {
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod().Name, "iPayroll");
            throw ex;
        }
    }

    public void ValidateSalaryGroup(SalaryGroupDTO _validate)
    {
        var query;
        try
        {
            if (_validate.CODE != null/* TODO Change to default(_) if this is not a reference type */ )
            {
                if (_validate.ID != 0)
                    query = (from p in Context.PA_SALARY_GROUP
                             where p.CODE.ToUpper == _validate.CODE.ToUpper
                                                          & p.ID != _validate.ID
                             select p).SingleOrDefault;
                else
                    query = (from p in Context.PA_SALARY_GROUP
                             where p.CODE.ToUpper == _validate.CODE.ToUpper
                             select p).FirstOrDefault;
                return (query == null);
            }
            return true;
        }
        catch (Exception ex)
        {
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod().Name, "iPayroll");
            throw ex;
        }
    }

    public bool ModifySalaryGroup(SalaryGroupDTO objSalaryGroup, UserLog log, ref decimal gID)
    {
        PA_SALARY_GROUP objSalaryGroupData = new PA_SALARY_GROUP() { ID = objSalaryGroup.ID };
        try
        {
            Context.PA_SALARY_GROUP.Attach(objSalaryGroupData);
            objSalaryGroupData.CODE = objSalaryGroup.CODE.Trim;
            objSalaryGroupData.NAME = objSalaryGroup.NAME.Trim;
            objSalaryGroupData.EFFECT_DATE = objSalaryGroup.EFFECT_DATE;
            objSalaryGroupData.REMARK = objSalaryGroup.REMARK;

            Context.SaveChanges(log);
            gID = objSalaryGroupData.ID;
            return true;
        }
        catch (Exception ex)
        {
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod().Name, "iPayroll");
            throw ex;
        }
    }

    public bool DeleteSalaryGroup(List<decimal> lstID)
    {
        List<PA_SALARY_GROUP> lstSalaryGroupData;
        try
        {
            lstSalaryGroupData = (from p in Context.PA_SALARY_GROUP
                                  where lstID.Contains(p.ID)
                                  select p).ToList;

            for (var idx = 0; idx <= lstSalaryGroupData.Count - 1; idx++)
            {
                foreach (var item in lstSalaryGroupData[idx].PA_SALARY_LEVEL)
                {
                    foreach (var item1 in item.PA_SALARY_RANK)
                        Context.PA_SALARY_RANK.DeleteObject(item1);
                    Context.PA_SALARY_LEVEL.DeleteObject(item);
                }

                Context.PA_SALARY_GROUP.DeleteObject(lstSalaryGroupData[idx]);
            }
            Context.SaveChanges();
            return true;
        }
        catch (Exception ex)
        {
            WriteExceptionLog(ex, MethodBase.GetCurrentMethod().Name, "iPayroll");
            throw ex;
        }
    }
}
